api/exec_start.go

```diff
@@ -0,0 +1,217 @@
+	return
+}
+
+// DemuxFrame reads contents for frame from server multiplexed stdin/stdout/stderr/2nd error channel
+func DemuxFrame(r io.Reader, buffer []byte, length int) (frame []byte, err error) {
+	if len(buffer) < length {
+		buffer = append(buffer, make([]byte, length-len(buffer)+1)...)
+	}
+
+	n, err := io.ReadFull(r, buffer[0:length])
+	if err != nil {
+		return nil, nil
+	}
+	if n < length {
+		err = io.ErrUnexpectedEOF
+		return
+	}
+
+	return buffer[0:length], nil
+}
+
+// This is intended to be run as a goroutine, handling resizing for a container
+// or exec session.
+func (c *API) attachHandleResize(ctx context.Context, resizeChannel <-chan drivers.TerminalSize, sessionId string) {
+	for {
+		select {
+		case <-ctx.Done():
+			c.logger.Warn("Resize handler is done")
+			return
+		case size := <-resizeChannel:
+			c.logger.Debug("Resize terminal", "sessionId", sessionId, "height", size.Height, "width", size.Width)
+			rerr := c.ExecResize(ctx, sessionId, size.Height, size.Width)
+			if rerr != nil {
+				c.logger.Warn("failed to resize TTY", "err", rerr)
+			}
+		}
+	}
+}
+
+// ExecStartAndAttach starts and attaches to a given exec session.
+func (c *API) ExecStart(ctx context.Context, sessionID string, options ExecStartRequest) error {
+	client := *c.httpClient
+	client.Timeout = 0
+
+	var socket net.Conn
+	socketSet := false
+	dialContext := client.Transport.(*http.Transport).DialContext
+	t := &http.Transport{
+		DialContext: func(ctx context.Context, network, address string) (net.Conn, error) {
+			c, err := dialContext(ctx, network, address)
+			if err != nil {
+				return nil, err
+			}
+			if !socketSet {
+				socket = c
+				socketSet = true
+			}
+			return c, err
+		},
+		IdleConnTimeout: time.Duration(0),
+	}
+	client.Transport = t
+
+	// Detach is always false.
+	// podman reference doc states that "true" is not supported
+	execStartReq := struct {
+		Detach bool `json:"Detach"`
+	}{
+		Detach: false,
+	}
+	jsonBytes, err := json.Marshal(execStartReq)
+	if err != nil {
+		return err
+	}
+	req, err := http.NewRequestWithContext(ctx, "POST", fmt.Sprintf("%s/exec/%s/start", c.baseUrl, sessionID), bytes.NewBuffer(jsonBytes))
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Content-Type", "application/json")
+	res, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+	if res.StatusCode != http.StatusOK {
+		return err
+	}
+
+	if options.Tty {
+		go c.attachHandleResize(ctx, options.ResizeCh, sessionID)
+	}
+
+	if options.AttachInput {
+		go func() {
+			_, err := io.Copy(socket, options.Stdin)
+			if err != nil {
+				c.logger.Error("Failed to send stdin to exec session", "err", err)
+			}
+		}()
+	}
+
+	defer func() {
+		c.logger.Debug("Finish exec session attach")
+	}()
+
+	buffer := make([]byte, 1024)
+	if options.Tty {
+		// If not multiplex'ed, read from server and write to stdout
+		_, err := io.Copy(options.Stdout, socket)
+		if err != nil {
+			return err
+		}
+	} else {
```

Comment:
Can we return nil after the copy and drop this else

Reply:
options.Tty is false for non interactive exec sessions, used by consul health checks. I tried to mimic what the docker driver does and the original podman code had this if/else case to support it.
We can try to simplify but it will require some testing